---
title: "Network Validation"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(data.table)
library(ggplot2)
library(cowplot)
library(ComplexHeatmap)
library(pbapply)
library(stringr)
library(circlize)
library(MineICA)
# library(igraph)

theme_set(theme_cowplot())


source("../scripts/notebooks_helpers/net_validation.R")
# source("../scripts/helpers/preprocess.R")
# source("../scripts/helpers/enrichment_analysis.R")
# source("../scripts/helpers/graphs.R")
# source("../scripts/helpers/heatmap.R")

knitr::opts_chunk$set(echo = TRUE)
memory.limit(32*1024)
```

## Load data

```{r}
geneid <- fread("../data/annotations/c_elegans.PRJNA13758.WS265.geneIDs.txt.gz")
geneid[V3 == "", V3 := V4]
geneid <- geneid[, .(GeneName = V2, GeneSymbol = V3, GeneType = V6)]
geneid <- geneid[! duplicated(GeneName)]
setkey(geneid, GeneName)
tissues_df <- fread("../data/tissue_unique_genes/genes_unique_to_tissues.csv.gz")

tissue_colors <- c("germ_line" = "black", 
                   "hypodermis" = "#5ac134", 
                   "intestine" = "#79bbe9", 
                   "muscle" = "#f68528",
                   "neurons" = "#E92128",
                   "multiple" = "#CCCCCC")
```

### Network

```{r}
net_data <- readRDS("../data/gene_network/all_genes_communities_d8_compact_file_inclusive.rds")

high_confidence_communities = readRDS("../data/gene_network/all_genes_communities_d8_high_confidence.rds")

if (0)
net_data$tissue_rho_list <- lapply(net_data$tissue_norm_list, function(x)
  cor(t(x), method = "s"))


net_data$tissue_communities_vector <- unlist(lapply(names(high_confidence_communities), function(i) {
  genes <- high_confidence_communities[[i]]
  setNames(rep_len(i, length(genes)), genes)
}))
```
```{r}
high_confidence_communities_flat = rbindlist(lapply(names(high_confidence_communities),function(x){
  r = data.frame(GeneName = high_confidence_communities[[x]])
  r$GeneSymbol = geneid[r$GeneName,GeneSymbol]
  r$Community  = x
  r
}))
```

```{r}
#rho_df <- rbindlist(lapply(names(net_data$filter_tissue_rho_list), function(i) {
  i = names(net_data$filter_tissue_rho_list)[1]
  genotype <- switch(i, n2 = "N2", daf2 = "daf-2(e1368)")
  
  rho <- net_data$filter_tissue_rho_list[[i]]
  df <- reshape2::melt(rho)
  df <- as.data.table(df)
  colnames(df) <- c("GeneName1", "GeneName2", "Correlation")
  
  df[, GeneName1 := as.character(GeneName1)]
  df[, GeneName2 := as.character(GeneName2)]
  
  df <- df[GeneName1 != GeneName2]
  
  df[, GeneSymbol1 := geneid[GeneName1]$GeneSymbol]
  df[, GeneSymbol2 := geneid[GeneName2]$GeneSymbol]
  df[, Genotype := genotype]
  
  df[, Community1 := net_data$tissue_communities_vector[GeneName1]]
  df[, Community2 := net_data$tissue_communities_vector[GeneName2]]
  
  df[is.na(Community1), Community1 := "Not in Community"]
  df[is.na(Community2), Community2 := "Not in Community"]
  
  rho_df = df
#}))
```

### RNAi

```{r}
if(0){
rnai_data <- readRDS("../data/formated_counts/counts_and_annots_net_validation.rds")
for (var in names(data)) {
  assign(var, data[[var]])
}
rm(data)
}
```


```{r}
if (0){
  da_models_df <- fread("../data/differential_analysis/models/net_validation.csv")
  
  # standard normalization
  files <- setdiff(list.files("../data/differential_analysis/tables/net_validation/"),
                   list.files("../data/differential_analysis/tables/net_validation/", pattern = "^ts_"))
  files = files[!grepl("uls60",files) & !grepl("daf2",files) ]
  
  da_df <- rbindlist(pblapply(files, function(filename) {
    out <- fread(paste0("../data/differential_analysis/tables/net_validation/", filename))
    colnames(out)[3:ncol(out)] <- c("log2FoldChange", "lfcSE", "pvalue", "padj")
    out[, Name := str_remove_all(filename, "[.]csv[.]gz")]
    out
  }))
  da_df <- merge(da_models_df[, .(Name, RNAi, Genotype)], da_df, by = "Name")
}

da_models_df <- fread("../data/differential_analysis/models/net_validation.csv")

# tissue_specific normalization
files <- setdiff(list.files("../data/tissue_specific_regression/tables/net_validation/"),
                 list.files("../data/tissue_specific_regression/tables/net_validation/", pattern = "^ts_"))

  files = files[!grepl("uls60",files) & !grepl("daf2",files) ]

da_df <- rbindlist(pblapply(files, function(filename) {
  out <- fread(paste0("../data/tissue_specific_regression/tables/net_validation/", filename))
  colnames(out)[12] = "log2FoldChange"
  colnames(out)[14] = "lfcSE"
  colnames(out)[16] = "pvalue"
  colnames(out)[18] = "padj"
  out = out[,.(filename,GeneName,GeneSymbol,Converged,ConvergenceCode,log2Disp,germ_line,soma,log2FoldChange, lfcSE, pvalue, padj)]
  out[, Name := str_remove_all(filename, "[.]csv[.]gz")]
  out
}))
#da_df$Day = as.integer(substring(da_df$Name,nchar(da_df$Name)))
#da_df$Day = as.integer(substring(da_df$Name,gregexpr(da_df$Name,"Set"))

da_df <- merge(da_models_df[, .(Name, RNAi, Genotype,Day)], da_df, by = c("Name"))


if (0){
# tissue-specific no  rmalization
files <- list.files("../data/differential_analysis/tables/net_validation/", pattern = "^ts_")

ts_da_df <- rbindlist(pblapply(files, function(filename) {
  out <- fread(paste0("../data/differential_analysis/tables/net_validation/", filename))
  colnames(out)[3:ncol(out)] <- c("log2FoldChange", "lfcSE", "pvalue", "padj", "Tissue")
  out[, Name := str_remove_all(filename, "[.]csv[.]gz")]
  out
}))
ts_da_df <- merge(da_models_df[, .(Name, RNAi, Genotype)], ts_da_df, by = "Name")
}
```


```{r}


high_confidence_communities_flat = rbindlist(lapply(names(high_confidence_communities),function(x){
  r = data.frame(GeneName = high_confidence_communities[[x]])
  r$GeneSymbol = geneid[r$GeneName,GeneSymbol]
  r$Community  = x
  r
}))
tested_RNAis = unique(da_df$RNAi)
high_confidence_communities_flat$tested = high_confidence_communities_flat$GeneSymbol %in%tested_RNAis
outside_communities = tested_RNAis[!tested_RNAis %in% high_confidence_communities_flat$GeneSymbol]

setkey(geneid,GeneSymbol)
high_confidence_communities_flat = rbind(high_confidence_communities_flat,
                                         data.table(GeneSymbol = outside_communities, 
                                                    GeneName = geneid[outside_communities,GeneName], 
                                                    Community  = rep("Not in Community",length(outside_communities)),
                                                    tested = rep(T,length(outside_communities))))
res = aggregate(GeneSymbol~Community,high_confidence_communities_flat[tested==T],FUN=length,drop=F)
res$Community_plot = substring(res$Community,10,)
res$Community_plot[res$Community_plot == "mmunity"] = "None"
ggplot(res,aes(Community_plot,GeneSymbol,label=GeneSymbol)) + geom_point() + geom_text_repel()+theme(axis.text.x = element_text(angle=90)) + xlab("Co-expression Group") + ylab("Number of\nmembers tested")
ggsave("../figures/03/targets_per_community.pdf",width=4,height=4)

```


```{r}
conditions_df <- unique(da_models_df[, .(RNAiName = geneid[RNAi, on = "GeneSymbol"]$GeneName, 
                                         RNAiSymbol = RNAi, 
                                         Genotype)])
conditions_df[, Community := net_data$tissue_communities_vector[as.character(RNAiName)]]
conditions_df[is.na(Community), Community := "Not in Community"]
conditions_df[, CommunityPlot := Community]
conditions_df[Community != "Not in Community", CommunityPlot := gsub("y", "y ", Community)]
conditions_df[, MeasuredCorrelation := RNAiName %in% rownames(net_data$filter_tissue_rho_list$n2)]

# sel_not_in_comm <- conditions_df[Community == "Not in Community" & MeasuredCorrelation == TRUE & Genotype == "N2"]$RNAiSymbol
sel_not_in_comm <- c("alh-10", "glc-1", "mig-6", "nlp-28", "srp-2", "fmo-5", "mak-1", "osm-11")
```


## Colors

```{r}
genotype_colors <- c(N2 = "#990000ff", `daf-2(e1368)` = "#0b5394ff")

day_colors <- c("goldenrod1", "darkgreen")

tissue_colors <- c("germ_line" = "black", 
                   "hypodermis" = "#5ac134", 
                   "intestine" = "#79bbe9", 
                   "muscle" = "#f68528",
                   "neurons" = "#E92128")

community_colors <- colorspace::rainbow_hcl(length(net_data$tissue_communities))
names(community_colors) <- paste0("Community", seq_along(community_colors))
```

```{r}

#setkey(geneid,GeneName)
#da_df[,GeneSymbol := geneid[GeneName,GeneSymbol]]
all_RNAi =  unique(da_df$RNAi)
rho_ts_da_df <- merge(rho_df[Genotype == "N2"], 
                      da_df[Genotype == "N2" & ! GeneSymbol %in% all_RNAi], 
                      by.x = c("GeneSymbol1","GeneName2", "Genotype"), 
                      by.y = c("RNAi", "GeneName", "Genotype"), 
                      allow.cartesian = TRUE)
rho_ts_da_df[, RNAi := GeneSymbol1]

rho_ts_da_df[, Community1Plot := Community1]
rho_ts_da_df[, Community2Plot := Community2]

rho_ts_da_df[Community1 != "Not in Community", Community1Plot := gsub("y", "y ", Community1)]
rho_ts_da_df[Community2 != "Not in Community", Community2Plot := gsub("y", "y ", Community2)]

sig_rho_ts_da_df <- rho_ts_da_df[ConvergenceCode == 0,]#[padj < 1]
setkey(tissues_df,GeneName)
sig_rho_ts_da_df$tissue = tissues_df[sig_rho_ts_da_df$GeneName2,Tissue]
sig_rho_ts_da_df[is.na(tissue),tissue:="multiple"]
sig_rho_ts_da_df[,tissue:=factor(tissue)]
```

## Effect on expression of genes in communities

```{r, fig.width=15, fig.height=15}
if (0){
ggplot(rho_ts_da_df, aes(padj, color = Community2Plot)) +
  scale_x_log10() +
  facet_wrap(Community1Plot ~ RNAi, scales = "free_x") +
  stat_ecdf() +
  theme(legend.position = "top") +
  xlab("Effect of RNAi (p-value)") +
  ylab("ECDF") +
  scale_color_manual(name = NULL, values = setNames(community_colors, gsub("y", "y ", names(community_colors))))
}
```

```{r, fig.width=15, fig.height=15}
if (0){
ggplot(sig_rho_ts_da_df, aes(abs(log2FoldChange), color = Community2Plot)) +
  facet_wrap(Community1Plot ~ RNAi, scales = "free_x") +
  stat_ecdf() +
  theme(legend.position = "top") +
  xlab("Effect of RNAi (log2 Scale)") +
  ylab("ECDF") +
  scale_color_manual(name = NULL, values = setNames(community_colors, gsub("y", "y ", names(community_colors))))
}
```

## Correlation between differential analysis and correlation?

```{r, fig.width=8, fig.height=8}
library(rsq)
corcor_df <- sig_rho_ts_da_df[, .(CorCorLFC = cor(Correlation, log2FoldChange)), by = .(Community1Plot, RNAi)]
corcor_df = corcor_df[order(abs(CorCorLFC)),]
sig_rho_ts_da_df$RNAi_f =factor(sig_rho_ts_da_df$RNAi,levels=corcor_df$RNAi)

corcor_ind_com_df <- sig_rho_ts_da_df[, .(CorCorLFC = cor(Correlation, log2FoldChange)), by = .(Community1Plot,Community2Plot, RNAi)]
corcor_ind_com_df$RNAi_f =factor(corcor_ind_com_df$RNAi,levels=corcor_df$RNAi)
corcor_ind_com_df = corcor_ind_com_df[order(abs(CorCorLFC)),]
ggplot(corcor_ind_com_df,aes(RNAi_f,CorCorLFC,color=Community2Plot))+geom_point()


best_RNAi = corcor_df[abs(CorCorLFC) > .25,RNAi]

ggplot(corcor_ind_com_df[RNAi %in% best_RNAi],aes(x=abs(CorCorLFC),color=Community2Plot))+stat_ecdf(lwd=1) +
  scale_color_manual(name = NULL, values = community_colors) + geom_vline(xintercept=0)+xlab("Correlation between Asych-Seq vs and RNAi Effet (magnitude)")
ggsave("../figures/03/S_community_correlation.pdf",width=8,height=8)

corcor_ind_tissue_df <- sig_rho_ts_da_df[, .(CorCorLFC = cor(Correlation, log2FoldChange)), by = .(Community1Plot,tissue, RNAi)]
corcor_ind_tissue_df$RNAi_f =factor(corcor_ind_tissue_df$RNAi,levels=corcor_df$RNAi)
corcor_ind_tissue_df = corcor_ind_tissue_df[order(abs(CorCorLFC)),]



ggplot(corcor_ind_tissue_df[RNAi %in% best_RNAi],aes(x=abs(CorCorLFC),color=tissue))+stat_ecdf(lwd=1) +
  scale_color_manual(name = NULL, values = tissue_colors) +
  geom_vline(xintercept=0)+xlab("Correlation between Asych-Seq vs and RNAi Effet (magnitude)")
ggsave("../figures/03/S_tissue_correlation.pdf",width=8,height=8)
```

#build models controlling for any residual tissue effect
```{r, fig.width=8, fig.height=8}
library(ggrepel)

if (0){ #linear regression
fits_testing_for_tissue_effect = lapply(unique(sig_rho_ts_da_df$RNAi),function(cur_RNAi){
  dat = sig_rho_ts_da_df[RNAi==cur_RNAi,]
  list(cross = lm(log2FoldChange~Correlation*tissue,dat), full=lm(log2FoldChange~Correlation+tissue,dat),reduced=lm(log2FoldChange~tissue,dat))
})

}
sig_rho_ts_da_df$FoldChange = 2^sig_rho_ts_da_df$log2FoldChange
fits_testing_for_tissue_effect = lapply(unique(sig_rho_ts_da_df$RNAi),function(cur_RNAi){
  dat = sig_rho_ts_da_df[RNAi==cur_RNAi,]
  list(cross = glm(FoldChange~Correlation*tissue,dat,family=gaussian(link="log")), 
       full=glm(FoldChange~Correlation+tissue,dat,family=gaussian(link="log")),
       reduced=glm(FoldChange~tissue,dat,family=gaussian(link="log")))
})

names(fits_testing_for_tissue_effect) = unique(sig_rho_ts_da_df$RNAi)

partial_rsq = rbindlist(lapply(names(fits_testing_for_tissue_effect),function(cur_RNAi){
  d = fits_testing_for_tissue_effect[[cur_RNAi]]
  dat = sig_rho_ts_da_df[RNAi==cur_RNAi,]
  print(cur_RNAi)
  data.frame(RNAi=cur_RNAi,rsq.p = rsq.partial(d[["full"]],d[["reduced"]])$partial.rsq, r.sq.full = with(summary(d$full), 1 - deviance/null.deviance), r.sq.reduced = with(summary(d$reduced), 1 - deviance/null.deviance))
}))
ggplot(partial_rsq,aes(r.sq.full,rsq.p,color=RNAi,label=RNAi)) + geom_abline(intercept=0,slope=1)+geom_point() + 
  geom_text_repel()+theme(legend.position="none")
  xlab("R-squared of Full model") + ylab("Partial R-squared removing tissue-scale effects") + xlim(0,.2) + ylim(0,.2)
  
ggplot(partial_rsq,aes(r.sq.full,rsq.p/r.sq.full,color=RNAi,label=RNAi)) + geom_abline(intercept=1,slope=0,lty=2,color="dark gray")+geom_point() + 
  geom_text_repel()+theme(legend.position="none")+
  xlab("Full R-squared") + ylab("Partial R-squared / Full R-Squared") 
  
ggsave("../figures/03/S_partial_vs_full_rsquared_from_prediction_model.pdf",width=8,height=4)

```

#build models controlling for any specific community effect
```{r, fig.width=16, fig.height=16}
library(ggrepel)

#sig_rho_ts_da_df$FoldChange = 2^sig_rho_ts_da_df$log2FoldChange
fits_testing_for_community_effect = lapply(unique(sig_rho_ts_da_df$RNAi),function(cur_RNAi){
  dat = sig_rho_ts_da_df[RNAi==cur_RNAi,]
  list(cross = glm(FoldChange~Correlation*Community2,dat,family=gaussian(link="log")), 
       full=glm(FoldChange~Correlation+Community2,dat,family=gaussian(link="log")),
       reduced=glm(FoldChange~Community2,dat,family=gaussian(link="log")))
})

names(fits_testing_for_community_effect) = unique(sig_rho_ts_da_df$RNAi)

partial_rsq = rbindlist(lapply(names(fits_testing_for_community_effect),function(cur_RNAi){
  d = fits_testing_for_community_effect[[cur_RNAi]]
  dat = sig_rho_ts_da_df[RNAi==cur_RNAi,]
  print(cur_RNAi)
  data.frame(RNAi=cur_RNAi,rsq.p = rsq.partial(d[["full"]],d[["reduced"]])$partial.rsq, r.sq.full = with(summary(d$full), 1 - deviance/null.deviance), r.sq.reduced = with(summary(d$reduced), 1 - deviance/null.deviance))
}))
ggplot(partial_rsq,aes(r.sq.full,rsq.p/r.sq.full,color=RNAi,label=RNAi)) + geom_abline(intercept=1,slope=0,lty=2,col="dark gray")+geom_point() + 
  geom_text_repel()+theme(legend.position="none")+
  xlab("Full R-squared") + ylab("Partial R-squared / Full R-squared") 

ggsave("../figures/03/S_partial_vs_full_rsquared_from_prediction_model_communities.pdf",width=8,height=4)

commnunity_effect_sizes = rbindlist(lapply(names(fits_testing_for_community_effect),function(cur_RNAi){
  d = fits_testing_for_community_effect[[cur_RNAi]]
  dat = sig_rho_ts_da_df[RNAi==cur_RNAi,]
  print(cur_RNAi)
  r = data.frame(summary(d$full)$coefficients)
  r$name = rownames(r)
  r$name = ifelse(substring(r$name,0,4) == "Comm",substring(r$name,20),r$name)
  r$name[r$name == "mmunity"] = "none"
  r$RNAi =cur_RNAi
  r
}))
max_effect = aggregate(Estimate~RNAi,commnunity_effect_sizes,max)
max_effect = max_effect[order(max_effect$Estimate,decreasing=T),]
commnunity_effect_sizes$RNAi_f = factor(commnunity_effect_sizes$RNAi,levels=max_effect$RNAi)
community_effect_colors = c(Correlation="black",setNames(community_colors,c(as.character(1:16),"none")))
ggplot(commnunity_effect_sizes[name!="(Intercept)"],aes(name,Estimate,color=name)) + geom_point()+geom_errorbar(aes(ymin=Estimate-Std..Error,ymax=Estimate+Std..Error)) + facet_wrap(~RNAi_f,scale="free_y")+scale_color_manual(values=community_effect_colors)

commnunity_effect_sizes_to_print = commnunity_effect_sizes[,c("RNAi","name","Estimate","Std..Error","Pr...t..")]
names(commnunity_effect_sizes_to_print) = c("Knockdown Target","Coefficient Name","Estimate","Std..Error","p-value")
fwrite(commnunity_effect_sizes_to_print,"../tables/XX_community_contribution_to_predictions.csv")

```

```{r,fig.width=8,fig.height=8}
library(ggrastr)
library(forcats)
corcor_df = corcor_df[order(abs(CorCorLFC),decreasing=F)]
best_RNAi = corcor_df$RNAi[nrow(corcor_df)-0:8]
corcor_df$RNAi_f = factor(corcor_df$RNAi,levels=(corcor_df$RNAi))
tp = sig_rho_ts_da_df[RNAi %in% best_RNAi & abs(log2FoldChange) < 4]
tp$RNA_f = droplevels(factor(tp$RNAi,levels=(levels(corcor_df$RNAi_f))))
tp=tp[order(RNA_f,decreasing=T),]
corcor_df2 = corcor_df2[RNAi %in% best_RNAi]
ggplot(tp[,], aes(Correlation, log2FoldChange,grouping=RNAi_f)) +
  facet_wrap(~(factor(RNAi,levels=rev(levels(corcor_df$RNAi_f)))),nrow=3,ncol=3) +
  geom_point_rast(aes(color = Community2Plot), data=tp[Community2Plot=="Not in Community"],size = 1, alpha = .7,raster.dpi=600) +
  geom_point_rast(aes(color = Community2Plot), data=tp[Community2Plot!="Not in Community"], size = 1, alpha = .7,raster.dpi=600) +
  geom_smooth(color = "purple", formula = y ~ x, se = FALSE, method = "lm", size = 1) +
  ggpp::geom_text_npc(aes(npcx = "left", npcy = "top",
                          label = RNAi,size=10),data=corcor_df[RNAi %in% best_RNAi,]) +
  ggpp::geom_text_npc(aes(npcx = "right", npcy = "bottom",
                          label = paste0(round(100 * CorCorLFC, 1), "%"),size=10),data=corcor_df[RNAi %in% best_RNAi,]) +
  #geom_abline() +
  scale_color_manual(name = NULL, values = community_colors) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0, linetype = "dashed") +
  theme(strip.background = element_blank(), strip.text.x = element_blank(),text=element_text(size=10),axis.text=element_text(size=10)) +
  
  xlab("Correlation in Single Worm Data") +
  ylab("Effect of RNAi (log2 Scale)")
  ggsave("../figures/03/asynch_corr_vs_effect.pdf",width=8,height=8)
```


```{r,fig.width=8,fig.height=8}
tp = sig_rho_ts_da_df[!(RNAi %in% best_RNAi) & abs(log2FoldChange) < 4]
tp$RNA_f = droplevels(factor(tp$RNAi,levels=(levels(corcor_df$RNAi_f))))
tp=tp[order(RNA_f,decreasing=T),]
corcor_df2 = corcor_df[!RNAi %in% best_RNAi,]
ggplot(tp[sample(nrow(tp)/100),], aes(Correlation, log2FoldChange,grouping=RNAi_f)) +
  facet_wrap(~(factor(RNAi,levels=rev(levels(corcor_df$RNAi_f)))),nrow=8,ncol=10) +
  geom_point_rast(aes(color = Community2Plot), data=tp[Community2Plot=="Not in Community"],size = 1, alpha = .7,raster.dpi=600) +
  geom_point_rast(aes(color = Community2Plot), data=tp[Community2Plot!="Not in Community"], size = 1, alpha = .7,raster.dpi=600) +
  geom_smooth(color = "purple", formula = y ~ x, se = FALSE, method = "lm", size = 1) +
  ggpp::geom_text_npc(aes(npcx = "left", npcy = "top",
                          label = RNAi,size=10), 
                      corcor_df2) +
  ggpp::geom_text_npc(aes(npcx = "right", npcy = "bottom",
                          label = paste0(round(100 * CorCorLFC, 1), "%"),size=10),
                      corcor_df2) +
  #geom_abline() +
  scale_color_manual(name = NULL, values = community_colors) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0, linetype = "dashed") +
  theme(strip.background = element_blank(), strip.text.x = element_blank(),text=element_text(size=10),axis.text=element_text(size=10)) +
  
  xlab("Correlation in Single Worm Data") +
  ylab("Effect of RNAi (log2 Scale)")
  ggsave("../figures/03/S_asynch_corr_vs_effect.pdf",width=16,height=16)
```
```{r}
hits = commnunity_effect_sizes_to_print[`Knockdown Target` %in% corcor_df[abs(CorCorLFC)>.25,RNAi],]
avg_cont = aggregate(Estimate~`Coefficient Name`,hits,mean)
std_cont = aggregate(Estimate~`Coefficient Name`,hits,function(x)sd(x)/sqrt(length(x)))
rownames(std_cont) = std_cont$`Coefficient Name`
avg_cont = avg_cont[order(avg_cont$Estimate,decreasing=T),]
avg_cont$stderr = std_cont[avg_cont$`Coefficient Name`,"Estimate"]

avg_cont$`Coefficient Name Factor` = factor(avg_cont$`Coefficient Name`, levels = avg_cont$`Coefficient Name`)
ggplot(avg_cont[avg_cont$`Coefficient Name`!="(Intercept)",],aes(`Coefficient Name Factor`,Estimate))+geom_point()+geom_errorbar(aes(ymin=Estimate-stderr,ymax=Estimate+stderr)) + theme(axis.text.x = element_text(angle=90))+xlab("Linear Regression Coefficient") + ylab("Effect size")
ggsave("../figures/03/S_contribution_of_community_to_change.pdf",width=8,height=8)

```



#build models removing all communities 
```{r, fig.width=8, fig.height=8}

sig_rho_ts_da_df$FoldChange = 2^sig_rho_ts_da_df$log2FoldChange
fits_testing_for_tissue_effect = lapply(unique(sig_rho_ts_da_df$RNAi),function(cur_RNAi){
  dat = sig_rho_ts_da_df[RNAi==cur_RNAi &,]
  list(cross = glm(FoldChange~Correlation*tissue,dat,family=gaussian(link="log")), 
       full=glm(FoldChange~Correlation+tissue,dat,family=gaussian(link="log")),
       reduced=glm(FoldChange~tissue,dat,family=gaussian(link="log")))
})

names(fits_testing_for_tissue_effect) = unique(sig_rho_ts_da_df$RNAi)

partial_rsq = rbindlist(lapply(names(fits_testing_for_tissue_effect),function(cur_RNAi){
  d = fits_testing_for_tissue_effect[[cur_RNAi]]
  dat = sig_rho_ts_da_df[RNAi==cur_RNAi,]
  print(cur_RNAi)
  data.frame(RNAi=cur_RNAi,rsq.p = rsq.partial(d[["full"]],d[["reduced"]])$partial.rsq, r.sq.full = with(summary(d$full), 1 - deviance/null.deviance), r.sq.reduced = with(summary(d$reduced), 1 - deviance/null.deviance))
}))
ggplot(partial_rsq,aes(r.sq.full,rsq.p,color=RNAi,label=RNAi)) + geom_abline(intercept=0,slope=1)+geom_point() + 
  geom_text_repel()+theme(legend.position="none")
  xlab("R-squared of Full model") + ylab("Partial R-squared removing tissue-scale effects") + xlim(0,.2) + ylim(0,.2)

ggsave("../figures/03/S_partial_vs_full_rsquared_from_prediction_model.pdf",width=8,height=8)


if (0){
to_plot ="ccch-1"
to_plot ="elo-5"
ggplot(sig_rho_ts_da_df[RNAi %in% to_plot& tissue != "multiple"], aes(Correlation, log2FoldChange)) +
  geom_point(aes(color = tissue), size = 1, alpha = .7) +
  geom_smooth(color = "purple", formula = y ~ x, se = FALSE, method = "lm", size = 1) +
 # ggpp::geom_text_npc(aes(npcx = "left", npcy = "bottom",
  #                        label = paste0(round(100 * CorCorLFC, 1), "%")), 
   #                   corcor_df) +
  #geom_abline() +
  scale_color_manual(name = NULL, values = tissue_colors) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0, linetype = "dashed") +
  theme(legend.position = "none") +
  xlab("Correlation in Single Worm Data") +
  ylab("Effect of RNAi (log2 Scale)")
}

```

#compare to genes associated with cohort scaling model scale factor
#use table of correlations (this isn't the best way to do it)
```{r}
scaling_genes = fread("../data/scaling_models/genes_that_correlate_with_scale_factor.csv.gz")

res = merge(high_confidence_communities_flat,scaling_genes,by="GeneName")
avg = aggregate(cor~Community,res,mean)
res$transformed = (res$cor+1)/2 #map correlation values between -1 and 1 to values between 0 and 1.
res = glm(transformed~Community,res,family=quasi(link="logit",variance = "mu(1-mu)"))
r = data.frame(summary(res)$coefficients);
r$Community = rownames(r)
r = data.table(r)
r$Community = substring(r$Community,19,)
r[Community == "mmunity",Community:="None"]
r[Community == "",Community:="(Intercept)"]
ggplot(r[Community != "(Intercept)"],aes(Community,Estimate)) + geom_hline(yintercept=0,lty=2,col="dark gray")+geom_point() + geom_errorbar((aes(ymin=Estimate-`Std..Error`,ymax=Estimate+`Std..Error`))) 

library(betareg)
```

```{r}
run_analysis_on_tissue_corrected_counts = T
#if (!run_analysis_on_tissue_corrected_counts){
source("../scripts/helpers/preprocess.R")
consensus_scale_factors = fread("../data/csm_scale_factors.csv.gz")
models = read.csv("../data/differential_analysis/models/single_and_pooled_worms.csv");
model_names = c("n2_series_all","d8_sw");

#model_names = c("d_glp1_m");
#model_names = c("d8_20v25C")
#model_names = c("ts_d_25C");
data <- readRDS("../data/formated_counts/counts_and_annots_single_and_pooled_worms.rds")
bc <- readRDS("../data/batch_corrected_counts/single_and_pooled_worms.rds")


counts =lapply(model_names, function(x){
                                  r = bc$batch_counts_list[[x]]
                                  r[,colnames(r) %in% data$annots$Sample]
                                 })
names(counts) = model_names

annots <- lapply(model_names,function(x){
  print(x)
                  r = data$annots[Strain=="QZ0" & Exclude==F & Sample %in% colnames(counts[[x]]),]
                  r$model = x;
                  r
                })
names(annots) = model_names;

counts =lapply(model_names, function(x){
    counts[[x]][,colnames(counts[[x]]) %in% annots[[x]]$Sample]
})
names(counts) = model_names

nf <- lapply(model_names,function(x){print(x);
  normalizationFactor(counts[[x]], groups = annots[[x]][,Strain])
  })
names(nf) = model_names;

norm <- lapply(model_names,function(x){normalizeCounts(counts[[x]], nf[[x]])})
names(norm) = model_names;

rm(data, bc)
#load in tissue-specific regresison
  #tc = fread("../data/tissue_specific_regression/tables/single_and_pooled_worms_2/n2_d8_sw.csv.gz")
  tc = readRDS("../data/tissue_specific_regression/rds/single_and_pooled_worms.rds")
  tissue_norm = list("n2_d8_sw_tissue_norm" = tc[["n2_d8_sw"]]$norm)
  tissue_norm_annots = list("n2_d8_sw_sw_tissue_norm" = tc[["n2_d8_sw"]]$annots)
  rm(tc)
  
    tc = readRDS("../data/tissue_specific_regression/rds/single_and_pooled_worms_n2_series_all.rds")
  tissue_norm[["n2_series"]] = tc[["n2_series"]]$norm
  tissue_norm_annots[["n2_series"]] = tc[["n2_series"]]$annots
  rm(tc)
#}
```
#estimate community cohesion at each timepoint
```{r}

dat = tissue_norm[["n2_series"]]
cur_annots = tissue_norm_annots[["n2_series"]]
days = unique(cur_annots$Day)
communiy_cohesion_timeseries = rbindlist(lapply(days,function(cur_day){
    print(cur_day)
    x = dat[,cur_annots[Day == cur_day & grepl("TS_QZ0",Sample),Sample]]
    rbindlist(lapply(names(high_confidence_communities),function(cur_com){
      print(cur_com)
      genes = high_confidence_communities[[cur_com]]
      gene_indicies = which(rownames(x) %in% genes)
      print(paste(length(gene_indicies), "genes"))
      flush.console()
      cohesion = 0
      count = 0;
      for (i in 1:length(gene_indicies)){
  			for (j in i:length(gene_indicies)){
  			  if (i==j) next;
  				cohesion = cohesion + abs(cor(x[gene_indicies[i],],x[gene_indicies[j],], method = "s"))
  				count = count+1
  			}
      }
      data.table(day=cur_day,community=cur_com,cohesion=cohesion/count)
		}))
	}))
```
#plot cohesion timeseries
```{r,fig.width=12,fig.height=6}

community_colors <- colorspace::rainbow_hcl(length(high_confidence_communities))
names(community_colors) <- paste0("Community", seq_along(community_colors))


lm_res=  rbindlist(lapply(unique(communiy_cohesion_timeseries$community),function(cur_com){
   x = communiy_cohesion_timeseries[community==cur_com,]
   x$day = as.numeric(as.character(x$day))
   r = data.frame(summary(lm(cohesion~day,x))$coefficients)
   r$name = rownames(r);
   r$community = cur_com;
   data.table(r)
}))

lm_res[,p_decoration := ""]
lm_res[Pr...t..<.05,p_decoration := "*"]
lm_res[Pr...t..<.01,p_decoration := "**"]
lm_res[Pr...t..<.001,p_decoration := "*"]
lm_res_just_day_effect = lm_res[name=="day"]
setkey(lm_res_just_day_effect,community)

lab_fun = function(x){paste0("G",substring(x,10),lm_res_just_day_effect[as.character(x),p_decoration])
}
lev = unique(communiy_cohesion_timeseries$community)
lev2 = substring(lev,10)
lev = lev[order(as.integer(lev2))]
communiy_cohesion_timeseries$community_f = factor(communiy_cohesion_timeseries$community,levels=lev)

gp1 = ggplot(communiy_cohesion_timeseries,aes(as.integer(as.character(day)),cohesion,color=community_f)) + geom_point() + geom_line() + scale_color_manual(values=community_colors)+facet_wrap(~community_f,labeller = labeller(community_f =lab_fun))+theme(strip.background = element_blank(),legend.position = "none")+scale_x_continuous(breaks=seq(2,12,by=2))+xlab("Age (Days)")+ylab("Group Cohesion")

library(ggrepel)
res = dcast(lm_res,community~name,value.var = "Estimate")
res2 = dcast(lm_res,community~name,value.var = "Std..Error")
res3 = dcast(lm_res,community~name,value.var = "p_decoration")
names(res3)[2:3] = paste0(names(res3),".p_decoration")[2:3]
res_m=merge(res,res2,by="community",suffixes=c(".val",".stderr"))
res_m=merge(res_m,res3,by="community")
gp2 = ggplot(res_m,aes(`(Intercept).val`,day.val,label=paste(substring(community,10),day.p_decoration),bg=community,color=community))+
  scale_color_manual(values=community_colors)+
  scale_fill_manual(values=community_colors)+
  geom_errorbar(aes(ymin=day.val-day.stderr,ymax=day.val+day.stderr))+
  geom_errorbarh(aes(xmin=`(Intercept).val`-`(Intercept).stderr`,xmax=`(Intercept).val`+`(Intercept).stderr`)) + 
  geom_point(size=3,color="black",pch=21) +
  geom_text_repel(point.padding=4,nudge_x=.03, nudge_y=.003,min.segment.length=1)+
  xlab("Cohesion at Start of Adulthod")+
  ylab("Rate of cohesion change (per day)")+
  theme(legend.position="none")
ggpubr::ggarrange(gp1,gp2)
ggsave("../figures/03/S_group_cohesion_with_age.pdf",width=16,height=8)
```

```{r,fig.width=8,fig.height=8}
library(qs)

#we need to load the raw scaling data to obtain the individual sample names stored in the consensus scale factor .csv.gz file
ref = "d1_glp1_d8_sw_QZ0"
path="../data/scaling_models/all/"

scal_dat = qread(paste0(path,ref,"=all.qs"))
consensus_scale_factors_dat = consensus_scale_factors;
consensus_scale_factors_dat$Sample = names(scal_dat$dat$coef)

community_expression = rbindlist(lapply(1:2,function(x){
  if (x==1){
    cur_annots = tissue_norm_annots[["n2_d8_sw_tissue_norm"]]#Strain=="QZ0" & model=="d8_sw" & Exclude==F,]
    dat = tissue_norm[["n2_d8_sw_tissue_norm"]]#norm[["d8_sw"]][,cur_annots$Sample]
    processing = "tissue_effects_removed"
  
  }
  else if (x==2){
    processing = "tissue_effects_included" 
    
    cur_annots = annots[["d8_sw"]]#Strain=="QZ0" & model=="" & Exclude==F,]
    dat = norm[["d8_sw"]]#norm[["d8_sw"]][,cur_annots$Sample]
  }else stop("???")

rbindlist(lapply(colnames(dat),function(cur_sample){
  rbindlist(lapply(names(high_confidence_communities),function(cur_com){
    genes = high_confidence_communities[[cur_com]]
    gene_indicies = which(rownames(dat) %in% genes)
    data.frame(community = cur_com,Sample=cur_sample,expression = mean(dat[gene_indicies,cur_sample]),processing=processing)
  }))
}))
}))
###complete below to compare with and without tissue specific normalization


norm_exp= rbindlist(lapply(unique(community_expression$processing),function(cur_proc){
  rbindlist(lapply(names(high_confidence_communities),function(cur_com){
    x = community_expression[community==cur_com & processing == cur_proc,]
    data.frame(norm_exp = scale(x$expression)[,1],community=rep(cur_com,nrow(x)),Sample = x$Sample,processing=cur_proc)
  }))
}))
res = merge(community_expression,norm_exp,by=c("community","Sample","processing"))
res2 = merge(res,consensus_scale_factors_dat[,.(Sample,consensus_scale_factor)],by="Sample")

lm_res=  rbindlist(lapply(unique(community_expression$processing),function(cur_proc){
  rbindlist(lapply(names(high_confidence_communities),function(cur_com){
   x = res2[community==cur_com & processing == cur_proc ,]
   d = lm(norm_exp~consensus_scale_factor,x)
   r = data.frame(summary(d)$coefficients)
   r$name = rownames(r);
   r$community = cur_com;
   r$processing = cur_proc
   r$r.sq = summary(d)$r.squared
   data.table(r)
}))
}))
library(ggrepel)
lm_res = lm_res[name!="(Intercept)",]
lm_res[,p_decoration := ""]
lm_res[Pr...t..<.05,p_decoration := "*"]
lm_res[Pr...t..<.01,p_decoration := "**"]
lm_res[Pr...t..<.001,p_decoration := "***"]

community_colors <- colorspace::rainbow_hcl(length(high_confidence_communities))
names(community_colors) <- paste0("Community", seq_along(community_colors))


lm_res_wide_1 = dcast(lm_res,community+name~processing,value.var="Estimate")
lm_res_wide_2 = dcast(lm_res,community+name~processing,value.var="Std..Error")
lm_res_wide_3 = dcast(lm_res,community+name~processing,value.var="Pr...t..")
lm_res_wide_4 = dcast(lm_res,community+name~processing,value.var="r.sq")
lm_res_wide_a = data.table(merge(lm_res_wide_1,lm_res_wide_2,by=c("community","name"),suffixes=c(".Estimate",".Std..Error")))
lm_res_wide_b = data.table(merge(lm_res_wide_3,lm_res_wide_4,by=c("community","name"),suffixes=c(".Pr...t..",".r.sq")))
lm_res_wide = data.table(merge(lm_res_wide_a,lm_res_wide_b,by=c("community","name")))
lm_res_wide = lm_res_wide[name != "(Intercept)"]
rng =range(lm_res_wide[,.(tissue_effects_included.Estimate+tissue_effects_included.Std..Error,
                          tissue_effects_included.Estimate-tissue_effects_included.Std..Error,
                          tissue_effects_removed.Estimate+tissue_effects_removed.Std..Error,
                          tissue_effects_removed.Estimate-tissue_effects_removed.Std..Error)])

ggplot(lm_res_wide,aes(tissue_effects_included.Estimate,tissue_effects_removed.Estimate,color=community,label=substring(community,10))) + 
  geom_hline(yintercept=0,col="dark gray",lty=2)+
  geom_vline(xintercept=0,col="dark gray",lty=2)+
  geom_abline(intercept=0,slope=1,col="dark gray")+ 
  geom_smooth(method="lm",se=F,col="black")+ 
  geom_errorbar(aes(ymin = tissue_effects_removed.Estimate-tissue_effects_removed.Std..Error,ymax = tissue_effects_removed.Estimate+tissue_effects_removed.Std..Error))+
   geom_errorbarh(aes(xmin = tissue_effects_included.Estimate-tissue_effects_included.Std..Error,xmax = tissue_effects_included.Estimate+tissue_effects_included.Std..Error))+
  geom_point(size=4)+
  geom_text_repel(size=6)+
  xlim(rng)+ylim(rng)+
  scale_color_manual(values=community_colors) +
  xlab("Community correlation with Scale Factor")+
  ylab("Community correlation with Scale Factor\nafter tissue-composition correction")+
  theme(legend.position="none")
ggsave("../figures/04/correlation_between_communities_and_scaling_with_and_without_tissue_normalization.pdf",width=6,height=8)

rng2 = range(lm_res_wide[,.(tissue_effects_included.r.sq,
                            tissue_effects_removed.r.sq)])
ggplot(lm_res_wide,aes(tissue_effects_included.r.sq,tissue_effects_removed.r.sq,color=community,label=substring(community,10))) + 
  geom_hline(yintercept=0,col="dark gray",lty=2)+
  geom_vline(xintercept=0,col="dark gray",lty=2)+
  geom_abline(intercept=0,slope=1,col="dark gray")+ 
  geom_smooth(method="lm",se=F,col="black")+ 
  geom_point(size=4)+
  geom_text_repel(size=6)+
  xlim(rng2)+ylim(rng2)+
  scale_color_manual(values=community_colors) +
  xlab("Scale factor R-squared")+
  ylab("Scale factor R-squared\nafter tissue-composition correction")+
  theme(legend.position="none")
ggsave("../figures/04/r_squared_of_scaling_with_and_without_tissue_normalization.pdf",width=6,height=8)


lab_fun = function(x,cur_processing){lm_res[processing==cur_processing,][x,paste("CoExp G",substring(community_f,10),p_decoration)]}
lev = unique(res2$community)
lev2 = substring(lev,10)
lev = lev[order(as.integer(lev2))]
res2$community_f = factor(res2$community,levels=lev)
lm_res$community_f = factor(lm_res$community,levels=lev)
setkey(lm_res,community_f)

lab_fun_included = function(x)lab_fun(x,"tissue_effects_included")
lab_fun_removed = function(x)lab_fun(x,"tissue_effects_removed")
 
ggplot(res2[processing=="tissue_effects_included"],aes(consensus_scale_factor,expression,color=community)) + geom_point() + scale_y_continuous(trans="log2")+ geom_smooth(method="lm",color="black",se=F)+facet_wrap(~community_f,scale="free",labeller = labeller(community_f =lab_fun_included),ncol=8,nrow=2) +
  scale_color_manual(values=setNames(community_colors,gsub("y ","y",names(community_colors)))) + theme(strip.background = element_blank(),legend.position="none")+ ylab("Mean Expression") + xlab("Consensus Scale Factor")

ggsave("../figures/04/consensus_scaling_vs_communities_tissue_effects_included.pdf",width=19,height=4)
ggplot(res2[processing=="tissue_effects_removed"],aes(consensus_scale_factor,expression,color=community)) + geom_point() + scale_y_continuous(trans="log2")+ geom_smooth(method="lm",color="black",se=F)+facet_wrap(~community_f,scale="free",labeller = labeller(community_f =lab_fun_removed),ncol=8,nrow=2) +
  scale_color_manual(values=setNames(community_colors,gsub("y ","y",names(community_colors)))) + theme(strip.background = element_blank(),legend.position="none")+ ylab("Mean Expression") + xlab("Consensus Scale Factor")

ggsave("../figures/04/consensus_scaling_vs_communities_tissue_effects_removed.pdf",width=19,height=4)

lab_fun2 = function(x){
  x1 = lm_res[processing=="tissue_effects_removed"][x,]
  x2 = lm_res[processing=="tissue_effects_included"][x,]
  paste("CoExp G",
        substring(x1$community_f,10),
"(",
x2$p_decoration,
"/",
x1$p_decoration,
")")}

ggplot(res2,aes(consensus_scale_factor,expression,color=community)) + geom_point(data =res2[processing=="tissue_effects_included"],color="black",alpha=.7) +
  geom_point(data =res2[processing=="tissue_effects_removed"],alpha=.7) +
  scale_y_continuous(trans="log2")+ 
  geom_smooth(method="lm",color="black",se=F,data =res2[processing=="tissue_effects_included"])+
  geom_smooth(method="lm",color="red",se=F,data =res2[processing=="tissue_effects_removed"])+
  facet_wrap(~community_f,scale="free",labeller = labeller(community_f =lab_fun2),ncol=8,nrow=2) +
  scale_color_manual(values=setNames(community_colors,gsub("y ","y",names(community_colors)))) + theme(strip.background = element_blank(),legend.position="none")+ ylab("Mean Expression") + xlab("Consensus Scale Factor")
ggsave("../figures/04/consensus_scaling_vs_communities_both.pdf",width=19,height=4)

```


```{r}
library(ComplexHeatmap)
community_expression_wide = dcast(community_expression,Sample~community,value.var="expression")
community_expression_wide_mat = as.matrix(community_expression_wide[,grepl("Community",names(community_expression_wide)),with=F])
colnames(community_expression_wide_mat) = substring(colnames(community_expression_wide_mat),10)
rownames(community_expression_wide_mat) = community_expression_wide$Sample
com_cor = cor(community_expression_wide_mat,method="s")
m = as.dendrogram(hclust(as.dist(1-com_cor, "ward.D2")))
for (i in 1:2){
  if (i==1)
    pdf("../figures/03/S_community_abundance_correlation.pdf",width=8,height=8)
 draw(Heatmap(com_cor,cluster_rows = m, 
              name = "Correlation",
              column_title="Co-Expression Groups",
              row_title="Co-Expression Groups",
              cluster_columns = m ))

  if (i==1) dev.off()
}
```



```{r, fig.width=20, fig.height=12}
#zoom in on specific hit
for (i in 1:2){
  if (i == 1) pdf("nlp28_dcar1_focus.pdf",width=20,height=12);
sig_rho_ts_da_df = sig_rho_ts_da_df[GeneSymbol1 %in% c("nlp-28","dcar-1"),]
corcor_df <- sig_rho_ts_da_df[, .(CorCorLFC = cor(Correlation, log2FoldChange)), by = .(Community1Plot, RNAi)]

plot(ggplot(sig_rho_ts_da_df, aes(Correlation, log2FoldChange)) +
  geom_smooth(color = "gray", formula = y ~ x, se = FALSE, method = "lm", size = .5) +
  geom_point(aes(color = Community2Plot), size = 1, alpha = .7) +
  ggpp::geom_text_npc(aes(npcx = "left", npcy = "bottom",
                          label = paste0(round(100 * CorCorLFC, 1), "%")), 
                      corcor_df) +
  
  ggrepel::geom_text_repel(aes(Correlation,log2FoldChange,color = Community2Plot,label=GeneSymbol2),sig_rho_ts_da_df[abs(Correlation) > .4 & GeneSymbol1 == "dcar-1" |
                                                                                                                       Correlation < -.25 & GeneSymbol1 == "nlp-28" |
                                                                                                                       Correlation >.35 & GeneSymbol1 == "nlp-28" |
                                                                                                                     abs(log2FoldChange) > 1,],cex=3,fontface = "italic") + 
  
  facet_wrap( ~ RNAi, scales = "free") +
  scale_color_manual(name = NULL, values = setNames(community_colors, gsub("y", "y ", names(community_colors)))) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0, linetype = "dashed") +
  theme(legend.position = "top") +
  xlab("Asynch-Seq Correlation") +
  ylab("Effect of RNAi (log2 Fold-Change)"))
  if (i == 1) dev.off();
}
```

```{r, fig.width=7.5, fig.height=7.5}
ggdf <- sig_rho_ts_da_df[Genotype == "N2" & 1 | (Community1 != "Not in Community" | RNAi %in% sel_not_in_comm)]

ggdf$short_community = '';
not_in_com = grepl("Not",ggdf$Community1Plot)
ggdf$short_community[not_in_com] = "NG";
ggdf$short_community[!not_in_com] = paste0("G",substring(ggdf$Community1Plot[!not_in_com],11))

ggdf[, Wrap := paste0(RNAi, " (", short_community, ")")]
corcor_df <- ggdf[, .(CorCorLFC = cor(Correlation, log2FoldChange)), by = .(Wrap,RNAi)]
ggdf$CorCorLFC = 0;
invisible(lapply(corcor_df$Wrap,function(x){
  cr = corcor_df[Wrap == x,"CorCorLFC"]
  ggdf[Wrap == x,"CorCorLFC"] <<- cr
  c()
}))
ggdf[, Wrap := factor(Wrap, ggdf[! duplicated(Wrap)][order(CorCorLFC)]$Wrap)]
```


```{r, fig.width=7.5, fig.height=7.5}
com_colors = community_colors;
names(com_colors) =gsub("y", "y ", names(community_colors))
com_colors = c(com_colors,"#000000");
names(com_colors)[length(com_colors)]="Not in Community";
lapply(list(c(1,9),c(10,dim(corcor_df)[1])),function(rng){
  RNAi_to_plot = as.character(corcor_df[order(-(corcor_df$CorCorLFC)^2),RNAi][rng[1]:rng[2]])
  gp = ggplot(ggdf[RNAi %in% RNAi_to_plot,], 
         aes(Correlation, log2FoldChange)) +
    geom_smooth(color = "black", formula = y ~ x, se = FALSE, method = "lm", size = .5) +
  ggrastr::rasterise( 
    geom_point(aes(color = Community2Plot), cex = .75),dpi=600) +
    ggpp::geom_text_npc(aes(npcx = "left", npcy = "bottom",
                            label = round(CorCorLFC, 3)), 
                        corcor_df[RNAi %in% RNAi_to_plot]) + 
    ggpp::geom_text_npc(aes(npcx = "right", npcy = "top",
                            label = Wrap), 
                        corcor_df[RNAi %in% RNAi_to_plot],fontface="italic") +
    facet_wrap( ~ round(-CorCorLFC^2,2)+Wrap,scales="free_y",ncol=3,drop=T) +
    #geom_abline() +
    scale_color_manual(name = NULL, values = setNames(com_colors, gsub("y", "y ", names(community_colors)))) +
    geom_smooth(se=F) +
    geom_vline(xintercept = 0, linetype = "dashed",color="#CCCCCC") +
    geom_hline(yintercept = 0, linetype = "dashed",color="#CCCCCC") +
    theme(legend.position = "none",  strip.background = element_blank(),
    strip.text.x = element_blank()) +
    xlab("Asynch-seq correlation") +
    ylab("Effect of RNAi (log2 Scale)")
  plot(gp)
  ggsave(paste0("../figures/sup_figure_6/correlation_log2_fold_change_n2_",rng[1],".pdf"), width = 10, height = 7.5)
})
```

```{r, fig.width=4, fig.height=5}
 res = as.data.frame(do.call(rbind, lapply(unique(ggdf$RNAi),function(X){
  tmp = ggdf[as.character(X)]
  tmp$FoldChange = 2^(tmp$log2FoldChange)
  tmp = tmp[tmp$FoldChange<100,]#RNAi cross-reactivity (eg vha-3 and vha-11) lead to spurious super high fold changes
  res = glm(FoldChange~Correlation,tmp,family=gaussian(link="log"))
  if (summary(res)[["coefficients"]][2,1] > 10)
    browser()
 data.frame(RNAi=X,community=unique(tmp$Community1Plot)[1],b=summary(res)[["coefficients"]][2,1],pval_bonferroni=summary(res)[["coefficients"]][2,4]*nrow(ggdf),correlation = cor(tmp$Correlation,tmp$log2FoldChange, method = "s"))
})))

res$sig = res$pval_bonferroni < .05
table(res[,c("sig")])
res$short_community = '';
not_in_com = grepl("Not",res$community)
res$short_community[not_in_com] = "NG";
res$short_community[!not_in_com] = paste0("G",substring(res$community[!not_in_com],11))

com_colors = community_colors;
#names(com_colors) =gsub("y", "y ", names(community_colors))
com_colors = c(com_colors,"#000000");
names(com_colors)[length(com_colors)]="Not in Community";
ggplot() +
 geom_point(aes(correlation,-log10(pval_bonferroni),color=community),res)+
  geom_hline(yintercept=-log10(0.01),linetype = "dashed",color="#CCCCCC")+
  geom_vline(xintercept=0,linetype = "dashed",color="#CCCCCC")+
 scale_color_manual(name = NULL, values = com_colors)+
  xlim(-.5,.5)+
  ggrepel::geom_text_repel(aes(correlation,-log10(pval_bonferroni),color=community,label = paste0(RNAi,"(",short_community,")")), res,cex=3,fontface = "italic") + 
  theme(legend.position = "none") +
  ylab("p-value (-log10)") +
  xlab("Correlation between Asynch-Seq\nprediction and RNAi effect")
ggsave("../figures/03/correlation_log2_fold_change_n2_volcano.pdf", width = 5, height = 3)
table(res$pval_bonferroni < .01)

```
```{r}
dat = dcast(sig_rho_ts_da_df,GeneName2+RNAi~GeneName1,value.var=c("Correlation"),drop=F) 

lm(value ~ . , data = dat[,c(""))
```

```{r, fig.width=15, fig.height=15}
corcor_df <- sig_rho_ts_da_df[, .(CorCorLFC = cor(Correlation^2, abs(log2FoldChange))), by = .(Community1Plot, RNAi)]

ggplot(sig_rho_ts_da_df, aes(Correlation^2, abs(log2FoldChange))) +
  geom_smooth(color = "purple", formula = y ~ x, se = FALSE, method = "lm", size = .5) +
  geom_point(aes(color = Community2Plot), size = 1, alpha = .7) +
  ggpp::geom_text_npc(aes(npcx = "left", npcy = "bottom",
                          label = paste0(round(100 * CorCorLFC, 1), "%")), 
                      corcor_df) +
  facet_wrap(Community1Plot ~ RNAi, scales = "free") +
  geom_abline() +
  scale_color_manual(name = NULL, values = setNames(community_colors, gsub("y", "y ", names(community_colors)))) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0, linetype = "dashed") +
  xlab("Squared Correlation in Single Worm Data") +
  ylab("Effect of RNAi (Absolute log2 Scale)")
```

```{r, fig.width=15, fig.height=15}
corcor_df <- rho_ts_da_df[, .(CorCorSig = cor(Correlation^2, -log10(padj))), by = .(Community1Plot, RNAi)]

ggplot(sig_rho_ts_da_df, aes(Correlation^2, -log10(padj))) +
  geom_point(aes(color = Community2Plot), size = 1, alpha = .7) +
  ggpp::geom_text_npc(aes(npcx = "right", npcy = "top",
                          label = paste0(round(100 * CorCorSig, 1), "%")), 
                      corcor_df) +
  facet_wrap(Community1Plot ~ RNAi, scales = "free") +
  scale_color_manual(name = NULL, values = setNames(community_colors, gsub("y", "y ", names(community_colors)))) +
  theme(legend.position = "top") +
  xlab("Squared Correlation in Single Worm Data") +
  ylab("Effect of RNAi (Significance)")
```

## Correlation of RNAi with other genes; differentially expressed or not

```{r, fig.width=15, fig.height=15}
ggplot(rho_ts_da_df, aes(Correlation^2, color = padj < 1e-3)) +
  facet_wrap(Community1Plot ~ RNAi, scales = "free_x") +
  stat_ecdf() +
  theme(legend.position = "top") +
  xlab("Effect of RNAi (p-value)") +
  ylab("ECDF")
```

```{r}
ecdf_df <- rbindlist(pblapply(seq_len(nrow(conditions_df)), function(i) {
  cond_df <- conditions_df[i]
  
  da <- ts_da_df[RNAi == cond_df$RNAiSymbol & Genotype == cond_df$Genotype]
  da_genes <- da[padj < 1e-2]$GeneName
  
  rho_df[, Significant := GeneName1 %in% da_genes & GeneName2 %in% da_genes]
  rho_df[, Significant := ifelse(Significant, "True", "False")]
  rho_df[, Significant := factor(Significant, c("True", "False"))]
  
  ggdf <- rbind(
    rho_df[Significant == "True"],
    rho_df[Significant == "False"][sample(.N, size = 10000)]
  )
  ggdf <- cbind(ggdf[, .(Correlation, Significant)], cond_df)
  ggdf
  
}))
```

```{r, fig.width=16,fig.height=8}
ecdf_df[, Wrap := paste0(RNAiSymbol, " (", CommunityPlot, ")")]
ecdf_df[, Wrap := factor(Wrap, ecdf_df[! duplicated(Wrap)][order(Community, RNAiSymbol)]$Wrap)]

ggplot(ecdf_df[Genotype == "N2" & (Community != "Not in Community" | RNAiSymbol %in% sel_not_in_comm)], 
       aes(Correlation^2, color = Significant)) +
  stat_ecdf() +
  facet_wrap( ~ Wrap) +
  scale_color_manual(name = "Differentially Expressed (RNAi)",
                     values = setNames(genotype_colors, NULL)) +
  xlab("Squared Gene-Gene Correlation") +
  ylab("ECDF") +
  xlim(c(0, 1)) +
  theme(legend.position = "top")
ggsave("../figures/sup_figure_6/ecdf_diff_ana_n2.pdf", width = 16, height = 8)
```
```{r}
#set up data tables for clustering and ICA

  geneid$GeneSymbol_mangled= str_replace(geneid$GeneSymbol,"-",".")
	RNAi_GeneNames = geneid$GeneName[geneid$GeneSymbol %in% unique(ts_da_df$RNAi)]

  tmp = ts_da_df[Genotype=="N2",c("RNAi","GeneSymbol","GeneName","log2FoldChange")]
  fc = dcast(tmp,GeneName+GeneSymbol~RNAi,value.var="log2FoldChange")
  tmp = ts_da_df[Genotype=="N2",c("RNAi","GeneSymbol","GeneName","padj")]
  pv = dcast(tmp,GeneName+GeneSymbol~RNAi,value.var="padj")
  rownames(fc) = fc$GeneName
  rownames(pv) = pv$GeneName
  pv = pv[,!(colnames(pv) %in% c("GeneName","GeneSymbol"))]
  fc = fc[,!(colnames(fc) %in% c("GeneName","GeneSymbol"))]
  
  #get rid of NAs
	pv_ts = apply(pv,2,FUN = function(x){x[is.na(x)] = 1; return(x)})
	fc_ts = apply(fc,2,FUN = function(x){x[is.na(x)] = 0; return(x)})
	
	#get rid of NAs
	pv_ts = apply(pv_ts,2,FUN = function(x){x[is.na(x)] = 1; return(x)})
	fc_ts = apply(fc_ts,2,FUN = function(x){x[is.na(x)] = 0; return(x)})
	
	sig_in_any_sample = apply(pv_ts,1,FUN=function(x)return( any(x<.001) ) );
	fc_sig_ts = fc_ts[sig_in_any_sample,]
	pv_sig_ts = pv_ts[sig_in_any_sample,]
	
	to_keep = which(!(rownames(pv_sig_ts) %in% RNAi_GeneNames))
	fc_sig_ts = fc_sig_ts[to_keep,]
	pv_sig_ts = pv_sig_ts[to_keep,]
	
	cm  = melt(net_data$tissue_communities_vector)
	cm$GeneName = rownames(	cm)
  community_memberships = data.frame(GeneName = RNAi_GeneNames)
  community_memberships = merge(community_memberships,cm,by="GeneName",all.x=T,all.y=F)
  community_memberships$value[is.na(community_memberships$value)]="Community0";
	
	community_memberships =  merge(community_memberships,geneid,by="GeneName")
	tissue_dat = unique(ts_da_df[,c("GeneName","Tissue")]);
	community_memberships =  merge(community_memberships,tissue_dat,by="GeneName")
	names(community_memberships) = c("GeneName","Community","GeneSymbol","GeneSymbol_mangled","Tissue")
	community_memberships$com = as.integer(substring(community_memberships$Community,10,10))
	
	
	rownames(community_memberships) = community_memberships$GeneSymbol;
	
```


```{r}
n_ica = 6
#run ICA
RNAis_to_exclude = c();
#RNAis_to_exclude = c("rheb-1","idh-1","rps-22","fox-1","F41C3.5")
fc_sig_ts = fc_sig_ts[,!(colnames(ts_da_df) %in% RNAis_to_exclude)]
#fc_sig = fc_sig[,!(colnames(fc_sig) %in% RNAis_to_exclude)]

res_ts = runICA(X=fc_sig_ts,nbComp=n_ica, method = "JADE", maxit=100000)
saveRDS(res_ts,"wt_ICA.rds")

#res = runICA(X=fc_sig,nbComp=ncomp, method = "JADE", maxit=100000)
ic_colors=  brewer.pal(12,"Paired")[seq(2,12,by=2)]

#res_ts$A = res_ts$A[,c(1,3,2,6,4,5)]
res_ts$A = res_ts$A[,c(4,5,6,3,2,1)]
res_ts$S = res_ts$S[,c(4,5,6,3,2,1)]
```

```{r}
#set up everything for RNAi response clustering
#cluster RNA's based on raw correlation and make nice dendogram heatmap
	
	recalc_corr=F
	if (recalc_corr){
		corr_m = matrix(NA,nrow=dim(fc_sig_ts)[2],ncol=dim(fc_sig_ts)[2])
		rownames(corr_m) = colnames(corr_m) = colnames(fc_sig_ts)
		for (i in 1:dim(fc_sig_ts)[2]){
			for (j in 1:dim(fc_sig_ts)[2]){
				corr_m[i,j] = cor(fc_sig_ts[,i],fc_sig_ts[,j], method = "s")
			}
		}
	}
	#colfunc <- colorRampPalette(c("blue", "white","red"))
	#rampbreaks <- seq(-1, 1, length.out = 512+1)
	if (1)
	#col=c("green","purple","blue","orange","red","brown","light green","black")
	#col=c("green","purple","blue","orange","red","brown","light green","black")
	ngroups = 4;
	heatmap_group_colors = c("black",brewer.pal(ngroups-1,"Dark2"))
	clustered_dat =hclust(as.dist(1-corr_m), method = "complete")
	RNAi_dendogram = as.dendrogram(clustered_dat)
	
	RNAi_dendogram_groups = cutree(clustered_dat,k=ngroups)
	RNAi_dendogram = color_branches(RNAi_dendogram, clusters=RNAi_dendogram_groups[order.dendrogram(RNAi_dendogram)],col=heatmap_group_colors[1:ngroups],groupLabels=T)
	
	
	RNA_dend_group_community_match = data.frame(GeneSymbol = names(RNAi_dendogram_groups),
					group= RNAi_dendogram_groups,
					community = community_memberships[names(RNAi_dendogram_groups),"com"]
					)
					
	RNA_dend_group_community_match = merge(RNA_dend_group_community_match,geneid[,c("GeneSymbol","GeneName")],
						by.x="GeneSymbol",by.y="GeneSymbol",suffixes=c("",""),all.x=T)
						
	
	RNA_dend_group_community_match$community[is.na(RNA_dend_group_community_match$community)] = 0			
	RNA_dend_group_community_match$in_community = 	0+ RNA_dend_group_community_match$community > 0	
	RNA_dend_group_community_match$com_name = as.character(RNA_dend_group_community_match$community)
	
	#we set RNAis in no community to 0; we set RNAis that aren't even tissue specific (they *can't* be in communities!) to -1
	RNA_dend_group_community_match$com_name[RNA_dend_group_community_match$community == 0] = ""
	
	RNA_dend_group_community_match$tissue_specific = RNA_dend_group_community_match$GeneName %in% tissue_dat$GeneName
	RNA_dend_group_community_match = merge(RNA_dend_group_community_match,tissue_dat[,c("GeneName","Tissue")],by="GeneName",suffixes=c("",""),all.x=T)
	RNA_dend_group_community_match$Tissue[is.na(RNA_dend_group_community_match$Tissue)] = ""
	
	RNA_dend_group_community_match$community[RNA_dend_group_community_match$tissue_specific == F] = -1
	
	RNA_dend_group_community_match$tissue_name = toupper(substring(RNA_dend_group_community_match$Tissue,0,1))
	
	#RNA_dend_group_community_match$com_name = paste(RNA_dend_group_community_match$com_name,toupper(substring(RNA_dend_group_community_match$Tissue,0,1)))
	
	community_colors = c("#000000","#555555",colorspace::rainbow_hcl(max(RNA_dend_group_community_match$community)))
	RNA_dend_group_community_match$color =  community_colors[RNA_dend_group_community_match$community+2]
	
	
	tissue_colors_from_oli <- c("germ_line" = "black",
	                   "hypodermis" = "#5ac134",
	                   "intestine" = "#79bbe9",
	                   "muscle" = "#f68528",
                   "neurons" = "#E92128")
		
	tissue_colors = data.frame(tissue_name=c("","N","H","M","I","G"),tissue_color=c("black","#E92128","#5ac134","#f68528","#79bbe9","gray"))
	
	RNA_dend_group_community_match = merge(RNA_dend_group_community_match,tissue_colors,by="tissue_name",suffixes=c("",""),all.x=T)
	rownames(RNA_dend_group_community_match) = RNA_dend_group_community_match$GeneSymbol
	
	RNA_dend_group_community_match$plotted_knockdown_group = -1;
	RNA_dend_group_community_match$plotted_knockdown_group[RNA_dend_group_community_match$group == 2] = 1
	RNA_dend_group_community_match$plotted_knockdown_group[RNA_dend_group_community_match$group == 4] = 2
	RNA_dend_group_community_match$plotted_knockdown_group[RNA_dend_group_community_match$group == 3] = 3
	RNA_dend_group_community_match$plotted_knockdown_group[RNA_dend_group_community_match$group == 1] = 4
	match_counts = table(RNA_dend_group_community_match[,c("community","plotted_knockdown_group")])
	match_counts2 = table(RNA_dend_group_community_match[,c("in_community","plotted_knockdown_group")])

	
	
	#the dendogram and hclust assign different group IDs, so we need to match them up
	#cutree labels groups based on the alphabetical order of the RNAi label
	#the dendogram labels groups according to the leaf order in the tree of RNAi labels, incremental from top to bottom
	dend_grp_matchup = unique(RNAi_dendogram_groups[rownames(corr_m)[order.dendrogram(RNAi_dendogram)]])
	heatmap_group_id_lookup =  match(1:ngroups,dend_grp_matchup)	#given a hclust group X, its ID in the dendogram will be dend_grp_lookup[X];

	
	RNAi_group_colors =  heatmap_group_colors[heatmap_group_id_lookup[1:ngroups]]
```

```{r,fig.width=3,fig.height=5}
	#look at correspondences between rnai response groups and complex heatmaps
	
  match_counts3 = match_counts/apply(match_counts,1,FUN=sum)
  print(match_counts3)
  match_counts4 = as.data.frame(melt(match_counts,id.vars=c(as.character(1:4))))
  match_counts4$community = as.factor(match_counts4$community)
  levels(match_counts4$community) = c("NG MT","NT TS",paste0("G",as.character(c(1,2,3,5,6,7))))
 gp =  ggplot(match_counts4,aes(x= community, y = plotted_knockdown_group, fill = value)) +
  geom_tile() +
    
    geom_text(aes(label=value),col="white") +
    geom_text(aes(label=value),col="white") +
   ylab("RNAi Knockdown Group")+
   xlab("Co-Expression Group")+ 
   theme(legend.position = "none") 
 plot(gp)
  ggsave(paste0("../figures/main_figure_5/RNAi_community_overlap.pdf"), width = 3, height = 5)
	
```


```{r,fig.width=9,fig.height=7.5}
#plot heatmap
	
	col_fun = colorRamp2(c(-1,0, 1), c("blue","white", "red"))
	cm = ColorMapping(col_fun = col_fun)
	
clr = rep("black",length(rownames(corr_m)));
	clr[7] = "red"
	ra = rowAnnotation(foo = anno_text(RNA_dend_group_community_match[rownames(corr_m),"com_name"], 
		location = 0.5, just = "center",
		 gp = gpar(fontsize = 6, fill = RNA_dend_group_community_match[rownames(corr_m),"color"],
		 col = "white"),width = max_text_width("1")*2),
		 bar = anno_text(RNA_dend_group_community_match[rownames(corr_m),"tissue_name"], 
			location = 0.5, just = "center",
			 gp = gpar(fontsize = 6, col = RNA_dend_group_community_match[rownames(corr_m),"tissue_color"],
		 fill = "#555555"),width = max_text_width("1"))
	)
		 
	ca = columnAnnotation(foo = anno_text(RNA_dend_group_community_match[rownames(corr_m),"com_name"], 
						location = 0.5, just = "center",
						 gp = gpar(fontsize = 6, fill = RNA_dend_group_community_match[rownames(corr_m),"color"],
						 col = "white"),height = max_text_height("1")*2,rot = 90),
					 bar = anno_text(RNA_dend_group_community_match[rownames(corr_m),"tissue_name"], 
						location = 0.5, just = "center",
						 gp = gpar(fontsize = 6, col = RNA_dend_group_community_match[rownames(corr_m),"tissue_color"],
						 fill = "#555555"),
						 height =max_text_height("1"),rot = 90)
						)
		
		
		 
		

	gp = Heatmap(corr_m,cluster_rows=RNAi_dendogram,cluster_columns=RNAi_dendogram,col = cm,
		row_names_gp = gpar(col =  heatmap_group_colors[heatmap_group_id_lookup[RNAi_dendogram_groups[rownames(corr_m)]]],cex=.65),
		#row_names_gp = gpar(),row_dend_reorder = FALSE,
		heatmap_legend_param = list(title="Spearman\nCorrelation"),
		bottom_annotation =ca,
		right_annotation =ra,
		column_names_gp = gpar(col = RNAi_group_colors[RNAi_dendogram_groups[rownames(corr_m)]],cex=.65),
		row_split = ngroups, column_split = ngroups,row_title = "Knockdown\nGroup %s",column_title = "Knockdown\nGroup %s",
		row_gap = unit(0, "mm"),column_gap = unit(0, "mm"),border = TRUE,use_raster = TRUE,row_names_centered =T, column_names_centered =T) 
	
  plot(gp)
	pdf("../figures/main_figure_5/RNAi_similarty_heatmap2.pdf",width = 9, height = 7.5)
	plot(gp)
  dev.off()


```

```{r,fig.width=3.5,fig.height=6.5}
#plot ica heatmap (makes the ICA heatmap panel for the figure)
if (1){
	#col=c("green","purple","blue","orange","red","brown","light green","black")
	#col=c("green","purple","blue","orange","red","brown","light green","black")
	clustered_dat =hclust(dist(res_ts$A), method = "complete")
	row_dend = as.dendrogram(clustered_dat)

	#the dendogram and hclust assign different group IDs, so we need to match them up
	#cutree labels groups based on the alphabetical order of the RNAi label
	#the dendogram labels groups according to the leaf order in the tree of RNAi labels, incremental from top to bottom
	#dend_grp_matchup = unique(grps[rownames(res_ts$A)[order.dendrogram(row_dend)]])
	#dend_grp_lookup =  match(1:,dend_grp_matchup)	#given a hclust group X, its ID in the dendogram will be dend_grp_lookup[X];

	col_fun = colorRamp2(c(min(res_ts$A),0, max(res_ts$A)), c("purple","white", "green"))
	cm = ColorMapping(col_fun = col_fun)
	
	if (0) ra = rowAnnotation(foo = anno_text(RNA_dend_group_community_match[rownames(res_ts$A),"com_name"], 
		location = 0.5, just = "center",
		 gp = gpar(fontsize = 6, fill = RNA_dend_group_community_match[rownames(res_ts$A),"color"],
		 col = "white"),width = max_text_width("1")*1.2))
		 
		 
		 ra = rowAnnotation(foo = anno_text(RNA_dend_group_community_match[rownames(res_ts$A),"com_name"], 
		 		location = 0.5, just = "center",
		 		 gp = gpar(fontsize = 6, fill = RNA_dend_group_community_match[rownames(res_ts$A),"color"],
		 		 col = "white"),width = max_text_width("1")*1.1),
		 		 bar = anno_text(RNA_dend_group_community_match[rownames(res_ts$A),"tissue_name"], 
		 			location = 0.5, just = "center",
		 			 gp = gpar(fontsize = 6, col = RNA_dend_group_community_match[rownames(res_ts$A),"tissue_color"],
		 		 		fill = "#555555"),width = max_text_width("1")*1.1)
		 	)
		 

	colnames(res_ts$A) = paste(as.character(seq(1:n_ica)))
	
	gp = Heatmap(res_ts$A,cluster_rows=row_dend,cluster_columns = FALSE,col=cm,column_names_rot=0,
		row_names_gp = gpar(col =  RNAi_group_colors[RNAi_dendogram_groups[rownames(res_ts$A)]],cex=.65),
		column_names_gp = gpar(col =  ic_colors),
		right_annotation =ra,
		heatmap_legend_param = list(title="IC Contribution"),
		use_raster = TRUE,row_names_centered =T, column_names_centered =T)
	plot(gp)
	pdf("../figures/main_figure_5/IC_heatmap3.pdf",width = 3.5, height = 6.5)
	plot(gp)
  dev.off()

}
```
```{r}

#write ICA gene loadings to disk
	#geneids = RNAi_fold_change_results$geneid
	rownames(geneid) = geneid$GeneName

	
	feature_cutoff_ts = apply(abs(res_ts$S),2,FUN=function(x)quantile(x,.9))
	genes_contributing_to_ICA_ts= NULL
	for (i in 1:(dim(res_ts$S)[2])){
	  dat = data.frame(GeneName=rownames(res_ts$S),ICA.loading=res_ts$S[,i])
	  dat$change = "";
	  dat$change[dat$ICA.loading < -feature_cutoff_ts[i]] = "-";
	  dat$change[dat$ICA.loading > feature_cutoff_ts[i]] = "+";
		dat = dat[order(abs(dat$ICA.loading),decreasing=F),]
		dat$GeneSymbol = unlist(geneid[dat$GeneName,"GeneSymbol"])
		dat$ICA = i;
		genes_contributing_to_ICA_ts = rbind(dat,genes_contributing_to_ICA_ts)
	}
	#reverse
	genes_contributing_to_ICA_ts = genes_contributing_to_ICA_ts[dim(genes_contributing_to_ICA_ts)[1]:1,]
	write.csv(genes_contributing_to_ICA_ts,"..\\data\\ica\\IC_membership_ts.csv",row.names=F)

```

```{r}
library(data.table)

#run bootstrapping

#fancy bootstrapping ecdfs
#ica is run only on transcripts that showed a significant change in any condition
#but our null model needs to include all genes measured in the RNAi experiment, so we need to be careful when we select these sets.
correlationDistributionByICA <- function(ica, genes_in_RNAi_experiment,rho,nboot = 0, feature_cutoff_quantile = .90, quantiles = seq(0,1,.005)) {

  # match genes between ICA and Correlation matrix
  RNAi_single_worm_intersect_genes <- intersect(rownames(rho), genes_in_RNAi_experiment)
  rho <- rho[RNAi_single_worm_intersect_genes, RNAi_single_worm_intersect_genes]
  diag(rho) <- NA

  significant_RNAi_single_worm_intersect_genes = intersect(rownames(ica$S),genes_in_RNAi_experiment)
#  rho_sig <- rho[RNAi_single_worm_intersect_genes, RNAi_single_worm_intersect_genes]
 # browser()
  ica$S <- ica$S[significant_RNAi_single_worm_intersect_genes, ]

  # remove diagonal
 # rho_nodiag <- rho
 # diag(rho_nodiag) <- NA

  # cutoff
  feature_cutoff <- apply(ica$S, 2, function(x) quantile(abs(x), feature_cutoff_quantile))

  # iterate through ICAs or bootstrap iterations
  if (nboot == 0) {
    #get the ecdf for each ICA
    iter <- seq_len(ncol(ica$S))
  } else {
    #run a bunch of iterations to estimate ecdf of the null model
    iter <- seq_len(nboot)
  }
  ks_results = list()

  val = rbindlist(lapply(iter, function(i) {

    if (nboot == 0) {
      #we are not bootstrapping, so get the names of genes in the ICA i
	in_ica <- names(ica$S[abs(ica$S[, i]) > feature_cutoff[i], i])
	in_ica = in_ica[in_ica %in% rownames(rho)]
	not_in_ica <- rownames(rho)[! (rownames(rho) %in% in_ica)]
	#browser()
	vals_in_group <- na.omit(as.numeric(rho[in_ica, in_ica]))
	vals_out_group <- na.omit(as.numeric(rho[not_in_ica, not_in_ica]))

	df <- data.table(Index = i,

		     Value = c(vals_in_group, 
			       vals_out_group),

		     Group = c(rep_len("Both genes in IC", length(vals_in_group)),
		       rep_len("Unrelated to IC", length(vals_out_group))))

	ks_results[[i]] <<- ks.test(vals_in_group,vals_out_group,alternative="less",exact=T)
    } else {
      #we are bootstrapping, so get a set of genes from the null model for this bootstrap iteration
      #get the size of the set to use
      ngenes <- length(names(ica$S[abs(ica$S[, 1]) > feature_cutoff[1], 1]))
      in_ica <- sample(rownames(rho), size = ngenes)
      vals_in_group <- na.omit(as.numeric(rho[in_ica, in_ica]))
      df <- data.table(Index = i,

			     Value = c(vals_in_group),

			     Group = c(rep_len("Both genes in bootstrap",length(vals_in_group))))

	ks_res = NA;

    }

    df <- df[, .(Quantile = quantiles, 
		 AbsoluteValue = quantile(abs(Value), quantiles, na.rm = TRUE)),
	     by = .(Group, Index)]
  }))
  return(list(df = val,ks_results=ks_results))
}
ecdf_df_list <- correlationDistributionByICA(res_ts, rownames(fc_ts),net_data$tissue_rho_list$n2, nboot = 0)

random_ecdf_df_list <- correlationDistributionByICA(res_ts,rownames(fc_ts), net_data$tissue_rho_list$n2, nboot = 100)[["df"]] %>% 
  .[, .(Q05 = as.numeric(quantile(AbsoluteValue, 0.05)),
	Q50 = as.numeric(quantile(AbsoluteValue, 0.50)), 
	Q95 = as.numeric(quantile(AbsoluteValue, 0.95))), 
    by = .(Group, Quantile)]

ncomps = dim(res_ts$S)[2]
```
```{r}
rand_d = random_ecdf_df_list[Quantile==.5,]
  all_df = rbind(ecdf_df_list[["df"]][Index == 1],
              ecdf_df_list[["df"]][Index == 2],
              ecdf_df_list[["df"]][Index == 3],
              ecdf_df_list[["df"]][Index == 4],
              ecdf_df_list[["df"]][Index == 5],
              ecdf_df_list[["df"]][Index == 6],fill=T)
	  ggplot(all_df[Quantile==.5,],aes(Index, AbsoluteValue,color=Group)) +
	           geom_point()
	  
```


```{r,fig.width=7,fig.height=7}
plot_separately = T
	if(plot_separately){
	 
		gp_list <- lapply(seq_len(ncomps), function(i) {
			ggplot(random_ecdf_df_list, aes(Q50, Quantile)) +
			  geom_ribbon(aes(xmin = Q05, xmax = Q95), alpha = .3,color="dark gray") +
			  geom_line(aes(x = AbsoluteValue, color = Group), ecdf_df_list[["df"]][Index == i],size=1) +
			  scale_color_manual(values=c(ic_colors[i],"#000000")) +
			#  ggtitle(paste("ICA", i,"D=",round(ecdf_df_list[["ks_results"]][[i]]$statistic,3),"p=",signif(ecdf_df_list[["ks_results"]][[i]]$p.value,3))) +
			   theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
	panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
		    ggtitle(paste("IC",i))+
			  ylab("cdf") +
			  xlab("Asynch-Seq Correlation")
			}
		)
		gp <- ggpubr::ggarrange(plotlist = gp_list, common.legend = TRUE,ncol =3,nrow=3)
		gp <- ggpubr::annotate_figure(gp, top = paste("Number of Components:", ncomps))
		plot(gp)
  ggsave(paste0("../figures/main_figure_5/IC_enrichment.pdf"),plot=gp,width=4,height=7)
		
	}else{
	
		dat = ecdf_df_list[["df"]][ecdf_df_list[["df"]]$Group ==  "Both genes in IC",]
		gp = ggplot()+
			geom_line(data=random_ecdf_df_list, aes(x=Q50, y=Quantile)) +
			  geom_ribbon(data=random_ecdf_df_list,aes(xmin = Q05, xmax = Q95,y=Quantile), alpha = .3,color="dark gray")+
			
			  geom_line(data=dat,aes(x = AbsoluteValue, y=Quantile,color = factor(Index))) +
			  
			  scale_color_manual(values=ic_colors) +
			 
			   theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
		panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
			  ylab("ECDF") +
			  xlab("Absolute Single Worm Correlation")
			  
		plot(gp)
  ggsave(paste0("../figures/main_figure_5/IC_enrichment.pdf"),plot=gp,width=4,height=7)
	}
```


```{r}

#plot ecdfs comparing the single-worm correlation values from top hits in each ICA (used to make panels in figures)
if (0){
	feature_cutoff = apply(res_ts$S,2,genes_to_use = names(single_worm_data$tissues),
			FUN=function(x,genes_to_use){quantile(abs(x[names(x)%in%genes_to_use]),.8)})
	quantlies = seq(0,1,.05)
	layout(matrix(1:ncomp,nrow=2,byrow=T))
	
	
	ind = sample(nrow(rho_ts_tri)*ncol(rho_ts_tri))
			 
	shuffled_res = matrix(c(rho_ts_tri)[ind],nrow=nrow(rho_ts_tri),ncol=ncol(rho_ts_tri))
	rownames(shuffled_res) = colnames(shuffled_res) = rownames(rho_ts_tri)
			 
	par(mar=c(4,4,0,0))
	for (i in 1:ncomp){
			group_mem = names(res_ts$S[abs(res_ts$S[,i])>feature_cutoff[i],i]);
			tissue_specific_group_mem = group_mem[group_mem %in% names(single_worm_data$tissues)]
			not_group_mem = rownames(rho_ts_tri)[!(rownames(rho_ts_tri) %in% group_mem)]
			vals_in_group = rho_ts_tri[tissue_specific_group_mem,tissue_specific_group_mem]
			vals_into_group = rho_ts_tri[tissue_specific_group_mem,not_group_mem]
			vals_out_group = rho_ts_tri[not_group_mem,not_group_mem]
			vals_in_group = vals_in_group[!is.na(vals_in_group)]
			vals_into_group = vals_into_group[!is.na(vals_into_group)]
			vals_out_group = vals_out_group[!is.na(vals_out_group)]
			
			
			vals_in_group_shuffled = shuffled_res[tissue_specific_group_mem,tissue_specific_group_mem]
			
			
			q_out = quantile(abs(vals_out_group),quantiles,na.rm=T)
			q_in = quantile(abs(vals_in_group),quantiles,na.rm=T)
			q_into = quantile(abs(vals_into_group),quantiles,na.rm=T)
			
			q_in_shuffled = quantile(abs(vals_in_group_shuffled),quantiles,na.rm=T)
			
			plot(q_out,quantiles,xlim=range(q_out,q_in,q_into),col="black",type="l",lwd=1,bty="n",ylab="cdf",xlab="Correlation among Individual Worms")
			lines(q_in,quantiles,col=ic_colors[i],lwd=1)
			lines(q_into,quantiles,col=ic_colors[i],lwd=1,lty=2)
			lines(q_in_shuffled,quantiles,col="grey",lwd=1,lty=1)
			legend("bottomright",bty="n",title=paste0("IC ",i),legend=c(paste0("Unrelated to IC (N=",length(vals_out_group),")"),
										     paste0("One gene in IC(N=",length(vals_into_group),")"),
										     paste0("Both genes in IC (N=",length(vals_in_group),")")),
										     col=c("black",rep(ic_colors[i],2)),lty=c(1,1,2))
	}
}
```

